ofrak_core/ofrak/service/data_service.py

```diff
@@ -381,6 +362,75 @@ def get_children_with_boundaries_intersecting_range(self, r: Range) -> List[Data
 
         return [self._children[data_id] for data_id in intersecting_model_ids]
 
+    def get_children_affected_by_ranges(
+        self, patch_ranges: List[Range]
+    ) -> Iterable[Tuple[DataId, Range]]:
+        # Build a flat map of various types of points of interest, to be scanned in order of offset
+        points_of_interest: List[Tuple[int, int, Union[Range, Set[DataId]]]] = []
+
+        # int values represent the order that different types of points should be processed in
+        # (if they all had the same offset)
+        POINT_CHILDREN_ENDS = -2  # End of one or more children
+        POINT_RANGE_END = -1  # End of a patched range
+        POINT_ZERO_LENGTH_RANGE = 0  # Special case of patched range: it
+        POINT_RANGE_START = 1  # Start of a patched range
+        POINT_CHILDREN_STARTS = 2  # Start of one or more children
+
+        for r in patch_ranges:
+            if r.length() == 0:
+                points_of_interest.append((r.start, POINT_ZERO_LENGTH_RANGE, r))
+            else:
+                points_of_interest.append((r.start, POINT_RANGE_START, r))
+                points_of_interest.append((r.end, POINT_RANGE_END, r))
+
+        for waypoint in self.waypoints():
+            # Make sure no 0-length children (start and end at same waypoint) are counted
+            # These are never affected by a patch, UNLESS the patch is specifically to them
+            # That case is handled outside this function
+            # (the data ID the patch is "for" is always affected, so no need to check for it here)
+            points_of_interest.append(
+                (
+                    waypoint.offset,
+                    POINT_CHILDREN_ENDS,
+                    waypoint.models_ending.difference(waypoint.models_starting),
+                )
+            )
+            points_of_interest.append(
+                (
+                    waypoint.offset,
+                    POINT_CHILDREN_STARTS,
+                    waypoint.models_starting.difference(waypoint.models_ending),
+                )
+            )
+
+        points_of_interest.sort()
```

Comment:
Can you add a note here explaining how this sorting works, for readability?

From reading it, I'm assuming that it sorts by offset first, then point type? having a trail here will help future readers.

Reply:
Yes - I think I didn't add as many comments as I originally meant to. I'll add a few more.